flooddrake/slope_limiter.py

```diff
@@ -0,0 +1,114 @@
+from __future__ import division
+from __future__ import absolute_import
+
+from firedrake import *
+
+
+def SlopeLimiter(w, b, VCG):
+    """ Slope limiter the prevention of shocks. This is from Kuzmin (2011)
+
+        :param w: state vector
+
+        :param b: bed :class:`Function`
+
+        :param VCG: Continuous :class:`MixedFunctionSpace`
+
+
+    """
+
+    V = w.function_space()
+
+    if V.mesh().geometric_dimension() == 2:
+
+        # split functions
+        h, mu, mv = split(w)
+
+        # split function spaces
+        v, vu, vv = split(V)
+        v_cg, vu_cg, vv_cg = split(VCG)
+
+    if V.mesh().geometric_dimension() == 1:
+
+        # split functions
+        h, mu = split(w)
+
+        # split function spaces
+        v, vu = split(V)
+        v_cg, vu_cg = split(VCG)
+
+    # make a function for bedding
+    b_ = Function(v).project(b)
+
+    # get free surface depth function
+    v_func = Function(v).project(h + b_)
+
+    W = Function(V).assign(w)
+
+    # define the cg function to find maximum and minimum values for each vertex
+    v_max_cg = Function(v_cg).assign(-1000000)
+    v_min_cg = Function(v_cg).assign(1000000)
+
+    # cell average functions of depth and free surface
+    c = Function(FunctionSpace(v.mesh(), 'DG', 0)).project(v_func)
+
+    # Kernel for searching for min and maxes of cell averages
+    vert_max_min_kernel = """
+        #define MIN(X, Y) (((X) < (Y)) ? (X) : (Y))
+        #define MAX(X, Y) (((X) > (Y)) ? (X) : (Y))
+        for(int i=0;i<vert_cell_max.dofs;i++){
+        vert_cell_max[i][0]=MAX(vert_cell_max[i][0],cell_av[0][0]);
+        vert_cell_min[i][0]=MIN(vert_cell_min[i][0],cell_av[0][0]);
+        }
+        """
+
+    # slope limiting kernel
+    slope_limiter_kernel = """ float u_min=1000000.0, u_max=-10000000.0, EPSILON=0;
+        #define MIN(X, Y) (((X) < (Y)) ? (X) : (Y))
+        #define MAX(X, Y) (((X) > (Y)) ? (X) : (Y))
+        for(int i=0;i<vert_cell_min.dofs;i++){
+        u_min=MIN(vert_cell_min[i][0],u_min);
```

Comment:
OK, so when I said to indent C code, I didn't mean to add one more level of indentation uniformly, but to have an indentations that you should have when just writing C code, e.g. the content of blocks between curly braces should be indented etc.
```c
for (int i = 0; i < vert_cell_min.dofs; i++) {
    u_min = MIN(vert_cell_min[i][0], u_min);
    u_max = MAX(vert_cell_max[i][0], u_max);
}
```